internal/daemon/worker/session/session.go

```diff
@@ -22,26 +22,240 @@ var errMakeSessionCloseInfoNilCloseInfo = errors.New("nil closeInfo supplied to
 
 // ConnInfo defines the information about a connection attached to a session
 type ConnInfo struct {
-	Id         string
-	ConnCtx    context.Context
-	ConnCancel context.CancelFunc
-	Status     pbs.CONNECTIONSTATUS
-	CloseTime  time.Time
+	Id     string
+	Status pbs.CONNECTIONSTATUS
+
+	// The context.CancelFunc for the proxy connection.  Calling this function
+	// closes the proxy connection.
+	connCtxCancelFunc context.CancelFunc
+
+	// The time the controller has successfully reported that this connection is
+	// closed.
+	CloseTime time.Time
+}
+
+// Session is the local representation of a session.  After initial loading
+// the only values that will change will be the status (readable from
+// GetStatus()) and the Connections (GetLocalConnections()).
+type Session struct {
+	lock        sync.RWMutex
+	client      pbs.SessionServiceClient
+	connInfoMap map[string]*ConnInfo
+	resp        *pbs.LookupSessionResponse
+	status      pbs.SESSIONSTATUS
+	cert        *x509.Certificate
+}
+
+func newSess(client pbs.SessionServiceClient, resp *pbs.LookupSessionResponse) (*Session, error) {
+	if resp.GetExpiration().AsTime().Before(time.Now()) {
+		return nil, fmt.Errorf("session is expired")
+	}
+
+	parsedCert, err := x509.ParseCertificate(resp.GetAuthorization().GetCertificate())
+	if err != nil {
+		return nil, fmt.Errorf("error parsing session certificate: %w", err)
+	}
+
+	s := &Session{
+		client:      client,
+		connInfoMap: make(map[string]*ConnInfo),
+		resp:        resp,
+		status:      resp.GetStatus(),
+		cert:        parsedCert,
+	}
+	return s, nil
+}
+
+// ApplyLocalConnectionStatus set's a connection's status to the one provided.
+// If there is no connection with the provided id, an error is returned.
+func (s *Session) ApplyLocalConnectionStatus(connId string, status pbs.CONNECTIONSTATUS) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	// Update connection status if there are any connections in
+	// the request.
+	connInfo, ok := s.connInfoMap[connId]
+	if !ok {
+		return fmt.Errorf("could not find connection ID %q for session ID %q in local state",
+			connId,
+			s.GetId())
+	}
+
+	connInfo.Status = status
+	if connInfo.Status == pbs.CONNECTIONSTATUS_CONNECTIONSTATUS_CLOSED {
```

Comment:
Should this also call the cancel function? I realize it's called elsewhere, but any reason not to do it here too?